backend/src/Edna/Upload/Service.hs

```diff
@@ -53,12 +76,147 @@ measurementsToSummary =
+uploadFile proj methodology description fileName content = do
+  uploadFile' proj methodology description fileName content =<<
+    either throwM pure (parseExperimentXls content)
+
+uploadFile' ::
+  SqlId Project -> SqlId TestMethodology -> Text -> Text -> LByteString ->
+  FileContents -> Edna FileSummary
+uploadFile' projSqlId@(SqlId proj) methodSqlId@(SqlId method)
+  description fileName fileBytes fc = do
+    let projId = ProjectId $ fromIntegral proj
+    let methodId = TestMethodologyId $ fromIntegral method
+    runSelectReturningOne' (lookup_ (esProject ednaSchema) projId)
+      `whenNothingM_`
+      throwM (UEUnknownProject projSqlId)
+    runSelectReturningOne' (lookup_ (esTestMethodology ednaSchema) methodId)
+      `whenNothingM_`
+      throwM (UEUnknownTestMethodology methodSqlId)
+
+    transact $ insertAll (fromIntegral proj) (fromIntegral method)
+  where
+    insertAll :: HasCallStack => Word32 -> Word32 -> Edna FileSummary
+    insertAll projId methodId = do
+      let fileMeasurements = fcMeasurements fc
+      targetToId <- HM.fromList <$> mapM insertTarget (keys fileMeasurements)
+      let compounds = hashNub $ flip concatMap (elems fileMeasurements) $
+            \(TargetMeasurements targetMeasurements) -> keys targetMeasurements
+      compoundToId <- HM.fromList <$> mapM insertCompound compounds
+
+      expFileId <- insertExperimentFile projId methodId (fcMetadata fc)
+        description fileName fileBytes
+
+      forM_ (toPairs fileMeasurements) $
+        \(targetName, TargetMeasurements targetMeasurements) ->
+          forM_ (toPairs targetMeasurements) $ \(compoundName, measurements) -> do
+            let
+              getId :: HasCallStack => Text -> HashMap Text Word32 -> Word32
+              getId name =
+                fromMaybe (error $ "unexpected name: " <> name) .
+                view (at name)
+            let targetId = getId targetName targetToId
+            let compoundId = getId compoundName compoundToId
+            insertExperiment expFileId compoundId targetId measurements
+
+      measurementsToSummary fileMeasurements
+
+insertTarget :: Text -> Edna (Text, Word32)
+insertTarget targetName = (targetName,) . fromIntegral . unSqlId <$> do
+  runInsert' $ Pg.insert
+    (esTarget ednaSchema)
+    (insertExpressions [TargetRec default_ (val_ targetName) default_])
+    (Pg.onConflict (Pg.conflictingFields tName) Pg.onConflictDoNothing)
+  fromMaybe (error $ "added target not found: " <> targetName)  <$>
+    targetNameToId targetName
+
+insertCompound :: Text -> Edna (Text, Word32)
+insertCompound compoundName = (compoundName,) . fromIntegral . unSqlId <$> do
+  runInsert' $ Pg.insert
+    (esCompound ednaSchema)
+    (insertExpressions [CompoundRec default_ (val_ compoundName) default_ default_])
+    (Pg.onConflict (Pg.conflictingFields cName) Pg.onConflictDoNothing)
+  fromMaybe (error $ "added compound not found: " <> compoundName)  <$>
+    compoundNameToId compoundName
```

Comment:
No, I can't use `runInsertReturningOne` because it returns nothing if this compound already exists. We [discussed](https://serokell.slack.com/archives/C01F8AQSHGE/p1616340241019200) it yesterday.

Reply:
But what about `runInsertReturningList`?

Reply:
It will return empty list if nothing is updated and I will need to do `SELECT` after that. I can do the following: `INSERT RETURNING`, check whether anything was returned, and if not, call `SELECT`. This way when we add a new target/compound we will execute one command instead of two. But there will be more code in this case and I don't think we'll ever notice this optimization, it's very minor.